text/0121-config-service.md

```diff
@@ -0,0 +1,178 @@
+
+The user must then set up the config service. This can be done through the collector, which can be set up to expose an arbitrary configuration service implementation. Depending on implementation, this allows the collector to either act as a stand-alone configuration service, or as a bridge to remote configurations of the user's monitoring and tracing backend by 'translating' the monitoring backend's protocol to comply with the OpenTelemetry configuration protocol.
+
+## Internal details
+
+In the future, it is intended to add per-metric configuration. For example, this would allow the user to collect 5xx server error counts ever minute, and CPU usage statistics every 10 minutes. The remote configuration protocol was designed with this in mind, meaning that it includes more details than simply the metric collection period.
+
+Our remote configuration protocol will support this call:
+
+```
+service DynamicConfig {
+  rpc GetConfig (ConfigRequest) returns (ConfigResponse);
+}
+```
+
+A request to the config service will look like this:
+
+```
+message ConfigRequest{
+
+  // Required. The resource for which configuration should be returned.
+  opentelemetry.proto.resource.v1.Resource resource = 1;
+
+  // Optional. The value of ConfigResponse.fingerprint for the last configuration
+  // a resource received that was successfully applied.
+  bytes last_known_fingerprint = 2;
+}
+```
+
+While the response will look like this:
+
+```
+message ConfigResponse {
+
+  // Optional. The fingerprint associated with this ConfigResponse. Each change
+  // in configs yields a different fingerprint.
+  bytes fingerprint = 1;
+
+  // Dynamic configs specific to metrics
+  message MetricConfig {
+
+
+    // A Schedule is used to apply a particular scheduling configuration to
+    // a metric. If a metric name matches a schedule's patterns, then the metric
+    // adopts the configuration specified by the schedule.
+
+    message Schedule {
+
+      // A light-weight pattern that can match 1 or more
+      // metrics, for which this schedule will apply. The string is used to
+      // match against metric names. It should not exceed 100k characters.
+      message Pattern {
+        oneof match {
+          string equals = 1;       // matches the metric name exactly
+          string starts_with = 2;  // prefix-matches the metric name
+        }
+      }
+
+      // Metrics with names that match at least one rule in the inclusion_patterns are
+      // targeted by this schedule. Metrics that match at least one rule from the
+      // exclusion_patterns are not targeted for this schedule, even if they match an
+      // inclusion pattern.
+
+      // For this iteration, since we only want one Schedule that applies to all metrics,
+      // we will not check the inclusion_patterns and exclusion_patterns.
+      repeated Pattern inclusion_patterns = 1;
+      repeated Pattern exclusion_patterns = 2;
```

Comment:
nits: 
  1. because exclusion patterns have higher priority per the comment above, I would move it to the first position.
  2. I am not sure this grammar is expressive enough. Is there any prior art for this?

Reply:
1. Changed!
2. The choice to use exact match and starts with was made because they are more efficient to apply and clearer to read than more expressive alternatives, and have in practice been the only required pattern types for specifying metric names in Google's internal metric collection configuration protocol. However we are open to also adding a more expressive grammar such as regex or glob as a third option for a pattern type, so that monitoring backends can optionally choose to support a more expressive grammar without affecting efficiency for the simpler pattern types.